### PR TITLE
error when reading class file with unknown newer jdk version

### LINKED ISSUE #18573: Compiler should provide better error messages when using unsupported JDK version
Latest release of JDK 21 might be a time when more and more people might want to migrate to latest Java version. However, if we'd try to do so, we might get a very nasty error.

## Compiler version
3.3.0 (fixed in 3.3.1 for JDK 21, however the issue would still be present in future JDK's if more assumptions of classfile format are violated)

## Minimized code

```Scala 
import java.nio.file.StandardOpenOption

@main def Test = {
  val x =  StandardOpenOption.CREATE
}
``` 

## Output

```scala 
>scalac main.scala
error while loading StandardOpenOption,
class file /modules/java.base/java/nio/file/StandardOpenOption.class is broken, reading aborted with class java.lang.RuntimeException
bad constant pool index: 0 at pos: 1200
-- [E008] Not Found Error: main.scala:4:30 -------------------------------------
4 |  val x =  StandardOpenOption.CREATE
  |           ^^^^^^^^^^^^^^^^^^^^^^^^^
  |   value CREATE is not a member of object java.nio.file.StandardOpenOption
2 errors found
```

## Expectation

Compiler should provide better error messages when using unsupported JDK version, eg. what is the latest supported version and which version was found in the classfile along with a possible way to mitigate this issue, eg. `please try to use the latest version the Scala compiler`

### PR DESCRIPTION
**Pull request by bishabosha**
when reading a classfile causes a runtime exception, report the version of the classfile, and request that the user checks JDK compatibility.

Considerations:
- should we test this?

currently we have no process for testing latest JDK, currently we still only test with JDK 16 and 8

fixes #18573

### CODE REVIEW

**Comment by sjrd** [comment]
Can we somehow only report that clear error message *if* there is actually an exception thrown while reading the class file?

What does Scala 2 do in this situation?

**Comment by bishabosha** [comment]
> What does Scala 2 do in this situation?

From what I see there is no capturing of the class file version, and all runtime exceptions will be wrapped in the same "class file 'foo' is broken" error as dotty.

@lrytz @SethTisue what do you think?

I also think it would be ideal to not pre-emptively error, but append an addendum to any error that might occur.

File: compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala [hunk 1]
```diff
@@ -55,6 +55,22 @@ object ClassfileParser {
...
+    val minorVersion = in.nextChar.toInt
+    val majorVersion = in.nextChar.toInt
+    if ((majorVersion < JAVA_MAJOR_VERSION) ||
+        ((majorVersion == JAVA_MAJOR_VERSION) &&
+         (minorVersion < JAVA_MINOR_VERSION)))
+      throw new IOException(
+        s"class file '${classfile}' has unknown version $majorVersion.$minorVersion, should be at least $JAVA_MAJOR_VERSION.$JAVA_MINOR_VERSION")
+    if majorVersion > JAVA_LATEST_MAJOR_VERSION then
+      throw new IOException(
+        s"class file '${classfile}' has unknown version $majorVersion.$minorVersion, and was compiled by a newer JDK than supported by this Scala version, please update to a newer Scala version.")
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala:71]: We could also put in the error message which is the latest major supported, and should we reverse map that to the corresponding JDK version? (more work for @Kordyjan release procedure)

**Comment by lrytz** [comment]
Can you reproduce the error described in the ticket? It seems to work for me.

```
➜ sandbox sc3 -version
Scala compiler version 3.3.1 -- Copyright 2002-2023, LAMP/EPFL
➜ sandbox java -version
openjdk version "21" 2023-09-19
OpenJDK Runtime Environment (build 21+35-2513)
OpenJDK 64-Bit Server VM (build 21+35-2513, mixed mode, sharing)
➜ sandbox cat main.scala
import java.nio.file.StandardOpenOption

@main def Test = {
  val x =  StandardOpenOption.CREATE
}
➜ sandbox sc3 main.scala
```

**Comment by bishabosha** [comment]
> Can you reproduce the error described in the ticket? It seems to work for me.

3.3.1 is patched for jdk 21, it should break reliably with 3.3.0

**Comment by som-snytt** [comment]
This happens to me frequently, during casual testing. Spurious version mismatches, including tasty format of stale files I don't care about during casual testing, are a mild annoyance. Thanks for any mitigation or amelioration.

**Comment by bishabosha** [comment]
> Also we don't need to keep updating the version number as Java releases come out

In this case we can still say "class file 'foo' is broken... btw the class file version was 65.0, please check the JDK compatibility of your Scala version (3.x.y)."

**Comment by SethTisue** [comment]
> Kordyjan removed the backport:nominated label last week

Curious why? I would have thought this was a strong candidate for back porting.

As for the PR in general, I agree with Lukas's remarks.

**Review by SethTisue — COMMENTED** [review]
> Kordyjan removed the backport:nominated label last week

Curious why? I would have thought this was a strong candidate for backporting.

My observation on Discord and other community forums is that yes, this definitely actually happens to users quite often, and the users are often newcomers trying to run Scala for the first time. These newcomers often simply do not share our veterans’ awareness that it's important to aggressively update compiler and build tool versions if you want to use a recent JDK. It's rather common that their package manager gives them a brand new JDK but the learning materials they are using specify older Scala and/or sbt versions.

As for the PR in general, it looks good to me, and I agree with Lukas's remarks.

**Comment by bishabosha** [comment]
> Curious why? I would have thought this was a strong candidate for backporting.

I would guess because it makes a hard error when none would have existed before, so far LTS compatibility is basically "must be identical unless a flag is used"

**Comment by bishabosha** [comment]
@SethTisue with the latest change, we only report the JDK version along with the broken version message, this will produce something like:
```scala
error while loading String,
  class file /modules/java.base/java/lang/String.class is broken (version 65.0),
  please check the JDK compatibility of your Scala version (3.4.0-RC1-bin-SNAPSHOT-nonbootstrapped),
  reading aborted with class java.lang.RuntimeException:
  foo bar
```

**Review by sjrd — APPROVED** [review]

**Review by SethTisue — APPROVED** [review]

**Comment by bishabosha** [comment]
rebased

**Pull request merged**